Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+    public var timestampValue: Timestamp? {
+        guard case let .timestamp(t) = self else {
+            return nil
+        }
+        return t
+    }
+
+    /// Return this BSON as an `Int` if possible.
+    /// This will coerce non-integer numeric cases (e.g. `.double`) into an `Int` if such coercion would be lossless.
+    public func asInt() -> Int? {
+        switch self {
+        case let .int32(value):
+            return Int(value)
+        case let .int64(value):
+            return Int(exactly: value)
+        case let .double(value):
+            return Int(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `Int32` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into an `Int32` if such coercion would be lossless.
+    public func asInt32() -> Int32? {
+        switch self {
+        case let .int32(value):
+            return value
+        case let .int64(value):
+            return Int32(exactly: value)
+        case let .double(value):
+            return Int32(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `Int64` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into an `Int64` if such coercion would be lossless.
+    public func asInt64() -> Int64? {
+        switch self {
+        case let .int32(value):
+            return Int64(value)
+        case let .int64(value):
+            return value
+        case let .double(value):
+            return Int64(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as a `Double` if possible.
+    /// This will coerce numeric cases (e.g. `.decimal128`) into a `Double` if such coercion would be lossless.
+    public func asDouble() -> Double? {
+        switch self {
+        case let .double(d):
+            return d
+        default:
+            guard let intValue = self.asInt() else {
+                return nil
+            }
+            return Double(intValue)
+        }
+    }
+
+    /// Return this BSON as a `Decimal128` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into a `Decimal128` if such coercion would be lossless.
+    public func asDecimal128() -> Decimal128? {
+        switch self {
+        case let .decimal128(d):
+            return d
+        case let .int64(i):
+            return Decimal128(String(i))
+        case let .int32(i):
+            return Decimal128(String(i))
+        case let .double(d):
+            return Decimal128(String(d))
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `[T]` if possible, transforming each element via the provided closure.
+    /// If the closure returns nil, the value will be discarded from the resultant array.
+    public func asArray<T>(_ transform: (BSON) -> T?) -> [T]? {
```

Comment:
this could be named `map` since that's basically what it's doing, but I didn't want to imply that cases besides `.array` could be mapped. It could have been named `mapArray` to solve that concern, but I went with `asArray` to be consistent with other coercion methods.

Reply:
did you have a particular use case in mind for this method / is this something we expect users to do a lot or to do a lot ourselves?
just wondering why have this vs. just letting users do `myBSON.arrayValue?.map` or `myBSON.arrayValue?.compactMap` themselves.

Reply:
Honestly, its only benefit is that it's shorter. The reason it exists at all is because I originally was using the following:
```swift
public func asArray<T>(of: T.Type) -> [T]? {
    self.arrayValue?.compactMap { $0.bsonValue as? T }
}
```
I quickly realized that this method is not a good idea. I was doing things like `a.getArray(of: Int.self)` or `a.getArray(of: [Int32.self])` and getting `[]` back unknowingly. To fix that, I decided on the user passing a lambda instead, but I never really evaluated whether the method was still worth using. 

I use it quite frequently in the driver and the tests, but that doesn't really mean it's _useful_. I'm not sure if the slight increase in brevity is worth the loss in explicitness, anymore. What do you think?

Reply:
Okay, for now I'll just remove it. We can always add new helpers in the future as needed.